File: day2/solution.py
with open("input.txt") as f:
    input = f.read()

import enum

# X Rock
# Y Paper
# Z Scisors


class Shape(enum.Enum):
    ROCK = 1
    PAPER = 2
    SCISORS = 3


shapes = []


for line in input.splitlines():
    a, b = line.split(" ")

    if a == "A":
        left = Shape.ROCK
    elif a == "B":
        left = Shape.PAPER
    elif a == "C":
        left = Shape.SCISORS
    else:
        raise RuntimeError("a: " + a)

    if b == "X":
        right = Shape.ROCK
    elif b == "Y":
        right = Shape.PAPER
    elif b == "Z":
        right = Shape.SCISORS
    else:
        raise RuntimeError("b: " + b)

    shapes.append((left, right))


score = 0

for left_shape, right_shape in shapes:
    score += right_shape.value

    if left_shape == right_shape:
        score += 3
        continue

    if left_shape == Shape.ROCK and right_shape == Shape.PAPER:
        score += 6
        continue
    if left_shape == Shape.ROCK and right_shape == Shape.SCISORS:
        continue

    if left_shape == Shape.PAPER and right_shape == Shape.ROCK:
        continue
    if left_shape == Shape.PAPER and right_shape == Shape.SCISORS:
        score += 6
        continue

    if left_shape == Shape.SCISORS and right_shape == Shape.ROCK:
        score += 6
        continue
    if left_shape == Shape.SCISORS and right_shape == Shape.PAPER:
        continue

print(score)

File: day2/solution.hs
import Data.List (sort)
import System.IO (IOMode (ReadMode), hGetContents, openFile)

-- Rock     1
-- Paper    2
-- Scisors  3

data Shape = None | Rock | Paper | Scisors deriving (Enum, Eq)

getShapes :: [String] -> [(Shape, Shape)]
getShapes = map getShapes
  where
    getShapes :: String -> (Shape, Shape)
    getShapes line = (getShape (head line), getShape (last line))
    getShape :: Char -> Shape
    getShape char
      | char `elem` ['A', 'X'] = toEnum 1
      | char `elem` ['B', 'Y'] = toEnum 2
      | otherwise = toEnum 3

getScore :: (Shape, Shape) -> Int
getScore (a, b)
  | a == b = 3 + fromEnum b
  -- Rock first
  | a == Rock && b == Paper = 6 + fromEnum b
  | a == Rock && b == Scisors = fromEnum b
  -- Paper first
  | a == Paper && b == Rock = fromEnum b
  | a == Paper && b == Scisors = 6 + fromEnum b
  -- Scisors First
  | a == Scisors && b == Rock = 6 + fromEnum b
  | a == Scisors && b == Paper = fromEnum b
  | otherwise = 1000000000000

main :: IO ()
main = do
  file <- openFile "input.txt" ReadMode
  content <- hGetContents file
  let fileLines = lines content

  putStr $ "The score is: " ++ show (sum . map getScore $ getShapes fileLines)
